### ISSUE TITLE
Make `cache` support methods

### ISSUE THREAD

**Issue by brandonwillard** [dottxt-ai/outlines#1145]
### What behavior of the library made you think about the improvement?

Currently, the `cache` function doesn't truly support methods (see #1129).  Let's consider either making the `cache` interface private/internal, adding support for this (and admitting that it's a public interface/utility), or&mdash;at the very least&mdash;adding a warning to prevent confusion.

**Comment by brandonwillard** [comment]
I noticed from the helpful test case in #1129&mdash;and [the relevant `vllm` code](https://github.com/vllm-project/vllm/blob/551ce01078a655068e5ec3764d0a55ac744ea425/vllm/model_executor/guided_decoding/outlines_logits_processors.py#L91)&mdash;that the cache decorator is being directly applied to the function object, so we may not even be able to tell when a function is a method in order to give a warning.

In other words, there's probably not a general way to warn people that they could be doing something undesirable for caching.  Because of that, we might want to make the caching utilities private and leave it up to the user to determine how caching should work, since all we can really guarantee is that underlying objects are serializable in a way that can work for caching.

**Comment by Quexington** [comment]
I did some research here and it seems like the reason the test case in the PR is failing is because the function being pickled for the cache has a different closure environment.  The fact that `store == []` for the first cache check and `store == [a_1]` for the second cache check means that the pickle serializations are different as indeed they should be.  Consider this minimal example which _does_ pass:

```
store = []

def update_store(a):
    store.append(a)

def test_get_cache_from_class_method(test_cache):
    class DummyObject:
        @classmethod
        @test_cache
        def dummy_function(cls, a):
            update_store(a)

    DummyObject.dummy_function(1)
    DummyObject.dummy_function(1)

    assert len(store) == 1
```
since `store` is not acquired via closure, the environment for both invocations of the function is the same and therefore the cache is actually hit and the side effect is prevented the second time around.

Unfortunately, I can't identify why the relevant `vllm` code is suffering from this since it's not using any closures.  But the takeaway should be that actually cloudpickle does work for classmethods, perhaps even better than one might suppose.
